walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,80 @@
 
 package com.google.sps;
 
+import java.util.Arrays;
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    List<TimeRange> requiredEvents = findAttendeeEvents(events, request.getAttendees());
+    List<TimeRange> optionalEvents = findAttendeeEvents(events, request.getOptionalAttendees());
+
+    List<TimeRange> allAttendees = findOpenTimes(requiredEvents, optionalEvents, request);
+    List<TimeRange> onlyOptional = findOpenTimes(Arrays.asList(), optionalEvents, request);
+    List<TimeRange> onlyMandatory = findOpenTimes(requiredEvents, Arrays.asList(), request);
+
+    if (allAttendees.size() > 0) {
+        return allAttendees;
+    } else if (requiredEvents.size() > 0) {
+        return onlyMandatory;
+    } else {
+        return onlyOptional;
+    }
+  }
```

Comment:
Is this necessary? When you run findAttendeeEvents, if the attendeeList is empty, you will return an empty List already.

Reply:
If all attendees return an empty list but mandatory is not empty, then it should ignore optional

Reply:
Talked offline, since there is no plan to do the extra challenge, this is fine :)